### PR TITLE
Add itertools.chain.__setstate__

### PR DESCRIPTION
**Pull request by nonzerofloat**

### CODE REVIEW

**Comment by youknowone** [comment]
@nonzerofloat you may forgot to set `git config user.email`

**Comment by nonzerofloat** [comment]
Generally git do not allow commit without email registered. That's perhaps the commit's email is not consistent with my account's email. I have fixed.

Correct lint errors.

File: vm/src/stdlib/itertools.rs [hunk 1]
```diff
@@ -61,18 +61,49 @@ mod decl {
...
+            {
+                return Err(vm.new_type_error(String::from("Arguments must be iterators.")));
+            }
+            *self.iterables.write() = source.try_to_value(vm)?;
+            self.cur_idx.store(0);
+            *self.cached_iter.write() = if let Some(active) = active {
+                Some(active.clone().get_iter(vm)?)
+            } else {
+                None
+            };
```
Comment (youknowone) [inline vm/src/stdlib/itertools.rs:89]:
```suggestion
            *self.cached_iter.write() = active.map(|active| active.clone().get_iter(vm)).transpose()?;
```

File: vm/src/stdlib/itertools.rs [hunk 2]
```diff
@@ -61,18 +61,49 @@ mod decl {
...
+            Ok(())
+        }
     }
+
     impl IterNextIterable for PyItertoolsChain {}
     impl IterNext for PyItertoolsChain {
         fn next(zelf: &Py<Self>, vm: &VirtualMachine) -> PyResult<PyIterReturn> {
+            let iterables = zelf.iterables.read();
```
Comment (youknowone) [inline vm/src/stdlib/itertools.rs:97]: This change may cause dead-lock

Comment (nonzerofloat) [reply vm/src/stdlib/itertools.rs:97]: Can you give me more detailed explanation for deadlock? I suspect the place is suitable. Deadlock is unexpected.

Comment (youknowone) [reply vm/src/stdlib/itertools.rs:97]: Guess we have a chain with self-editing iterable on `__iter__`.  Previously, we locked iterable only before and after `get_iter`, it is safe.
After change, this line will hold the lock for entire loop. Any write operation in get_iter will be permanently blocked.

File: vm/src/stdlib/itertools.rs (hunk 1 above)
Comment (nonzerofloat) [reply vm/src/stdlib/itertools.rs:89]: Accepted

**Comment by rng-dynamics** [comment]
The implementation of `itertools.chain` has changed with #3788. I guess we should rebase this MR onto main (or something similar).

**Comment by youknowone** [comment]
Oh, this is done in #4232 

Thanks @dannasman !

File: vm/src/stdlib/itertools.rs [hunk 3]
```diff
@@ -58,7 +58,36 @@ mod decl {
         fn class_getitem(cls: PyTypeRef, args: PyObjectRef, vm: &VirtualMachine) -> PyGenericAlias {
             PyGenericAlias::new(cls, args, vm)
         }
+
+        #[pymethod(magic)]
+        fn setstate(&self, state: PyTupleRef, vm: &VirtualMachine) -> PyResult<()> {
+            let args = state.as_slice();
+            if args.is_empty() {
+                let msg = String::from("function takes at leat 1 arguments (0 given)");
```
Comment (Josh015) [inline vm/src/stdlib/itertools.rs:66]: `leat` should be `least`.

**Pull request closed without merging**